pkg/controllermanager/controller/controllerregistration/controllerregistration_seed_control.go

```
@@ -0,0 +1,458 @@
+		}
+	}
+
+	return controllerRegistrationNameToObject, kindTypeToControllerRegistrationName
+}
+
+// computeWantedControllerRegistrationNames computes the list of names of ControllerRegistration objects that are desired
+// to be installed. The computation is performed based on a list of required kind/type combinations and the proper mapping
+// to existing ControllerRegistration objects.
+func computeWantedControllerRegistrationNames(
+	wantedKindTypeCombinations sets.String,
+	kindTypeToControllerRegistrationName map[string]string,
+) (sets.String, error) {
+	names := sets.NewString()
+
+	for _, requiredExtension := range wantedKindTypeCombinations.UnsortedList() {
+		name, ok := kindTypeToControllerRegistrationName[requiredExtension]
+		if !ok {
+			return nil, fmt.Errorf("need to install an extension controller for %q but no appropriate ControllerRegistration found", requiredExtension)
+		}
+
+		names.Insert(name)
+	}
+
+	return names, nil
+}
+
+// computeRegistrationNameToInstallationNameMap computes a map that maps the name of a ControllerRegistration to the name of an
+// existing ControllerInstallation object that references this registration.
+func computeRegistrationNameToInstallationNameMap(
+	controllerInstallationList *gardencorev1beta1.ControllerInstallationList,
+	controllerRegistrationNameToObject map[string]*gardencorev1beta1.ControllerRegistration,
+	seedName string,
+) (map[string]string, error) {
+	registrationNameToInstallationName := make(map[string]string)
+
+	for _, controllerInstallation := range controllerInstallationList.Items {
+		if controllerInstallation.Spec.SeedRef.Name != seedName {
+			continue
+		}
+
+		if _, ok := controllerRegistrationNameToObject[controllerInstallation.Spec.RegistrationRef.Name]; !ok {
+			return nil, fmt.Errorf("ControllerRegistration %q does not exist", controllerInstallation.Spec.RegistrationRef.Name)
```

Not sure about this one.
If the referenced ControllerRegistration is (for some reason) not existing anymore, this will also block the whole controllerregistration reconciliation.

---

Yes, what would you expect? This is a system error that should never happen. Silently ignoring it and continuing with the rest would be wrong, I guess.

---

Agreed, this will anyways only happen if the operator removes the registration forcefully (i.e. manually removes the finalizer). So, I guess it's fine because the operator shouldn't need to do that in any case.